slides/week3/week3.org

```diff
@@ -0,0 +1,54 @@
+#+TITLE: Week 3
+#+AUTHOR: Kristaps Berzinch, Raymond Ortiz
+#+EMAIL: kristaps@robojackets.org, rortiz9@gatech.edu
+
+* TODO Needs speaker notes
+
+* What are we doing today?
+- Data structures and algorithms in C++
+- Brancing and merging in Git
+
+* What's a constructor?
+- A special type of function that instantiates a class
+- Basically, it sets up an object and returns it
+
+* What's a container?
+- An object you can use to store other objects
+- They can be ordered, associative, both, or neither
+  - *Ordered:* The contents are in a specific order
+  - *Associative:* Two objects are associated with each other within the container
+- TODO am I missing any important properties here? I want to say "uniqueness" or something...
+
+* Common containers
+| =array= | a static-length ordered list of objects |
+| =vector= | a dynamic-length ordered list of objects |
+| =set= | an ordered group of unique objects |
+| =map= | an ordered group of key-value pairs, with unique keys|
+
+* Arrays
+| =at= | returns the value at a specific index |
+| =front= | returns the first item |
+| =back= | returns the last item |
+| =size= | returns the number of elements |
+
+* Vectors
+- Everything from the last slide, and...
+| =clear= | removes all items |
+| =insert= | inserts an item at a specific index |
+| =push_back= | adds a given element to the end |
+| =pop_back= | removes the last element |
+
+* Iterators
+- An object that lets you access objects in a container
+- Depending on their type, they will let you read or write data in the container, and move forwards, backwards, or arbitrarily
+- The containers library includes the most common ones you'll need, but you can also write your own
+
+* Algorithms
+- All of these can be restricted to a range in the container
+| =sort= | sorts a container |
+| =count= | counts the number of items in a container that match a given item |
+| =find= | returns an iterator to the first element that matches a given item |
+| =copy= | copies elements from one container to another |
+| =fill= | fills a container with copies of a given element |
+| =accumulate= | sums all elements in a container |
```

Comment:
Transform (map) is very useful when paired with accumulate (reduce) (eg: transform to multiply all elements by 2, then accumulate to sum/subtract/multiply them together).